python/hail/table.py

```diff
@@ -1450,12 +1461,13 @@ def _process_joins(self, *exprs):
         broadcast_f = lambda left, data, jt: Table(left._jt.annotateGlobalJSON(data, jt))
         left, cleanup = process_joins(self, exprs, broadcast_f)
 
-        if left is not self:
+        if left is not self and (original_key is None and left.key is not None
+                                 or original_key is not None and (left.key is None or list(left.key) != original_key)):
             if original_key is not None:
                 if original_key != list(left.key):
                     left = left.key_by(*original_key)
             else:
-                left = left.key_by(None)
+                left = left.key_by()
```

Comment:
kind of disappointing we can't just do `key_by(original_key)`

Reply:
I concluded this is a bug keying at the end -- I've moved it into the join function.